be/src/exprs/math_functions.cpp

```diff
@@ -899,10 +898,310 @@ StatusOr<ColumnPtr> MathFunctions::cosine_similarity(FunctionContext* context, c
+
+    auto pattern = ColumnHelper::get_const_value<TYPE_VARCHAR>(base);
+
+    std::string pattern_str = pattern.to_string();
+    std::string delimiter = ",";
+    size_t pos_start = 0, pos_end, delim_len = delimiter.length();
+    std::string token;
+    while ((pos_end = pattern_str.find(delimiter, pos_start)) != std::string::npos) {
+        token = pattern_str.substr(pos_start, pos_end - pos_start);
+        CppType num = ::atof(token.c_str());
+        base_elements.push_back(num);
+        pos_start = pos_end + delim_len;
+    }
+    float last = ::atof(pattern_str.substr(pos_start).c_str());
+    base_elements.push_back(last);
+    size_t base_flat_size = base_elements.size();
+
+    const Column* target_flat = down_cast<const ArrayColumn*>(target)->elements_column().get();
+    const uint32_t* target_offset = down_cast<const ArrayColumn*>(target)->offsets().get_data().data();
+
+    if (target_flat->has_null()) {
+        return Status::InvalidArgument("cosine_similarity does not support null values");
+    }
+    if (target_flat->is_nullable()) {
+        target_flat = down_cast<const NullableColumn*>(target_flat)->data_column().get();
+    }
+    const CppType* target_data_head = down_cast<const ColumnType*>(target_flat)->get_data().data();
+    size_t target_size = target->size();
+
+    // prepare result with nullable value.
+    ColumnPtr result = ColumnHelper::create_column(TypeDescriptor{TYPE}, false, false, target_size);
+    ColumnType* data_result = down_cast<ColumnType*>(result.get());
+    CppType* result_data = data_result->get_data().data();
+
+    for (size_t i = 0; i < target_size; i++) {
+        size_t t_dim_size = target_offset[i + 1] - target_offset[i];
+        if (t_dim_size != base_flat_size) {
+            return Status::InvalidArgument(
+                    fmt::format("cosine_similarity requires equal length arrays in each row. base array dimension size "
+                                "is {}, target array dimension size is {}.",
+                                base_flat_size, t_dim_size));
+        }
+        if (t_dim_size == 0) {
+            return Status::InvalidArgument("cosine_similarity requires non-empty arrays in each row");
+        }
+    }
+
+    const CppType* target_data = target_data_head;
+    const CppType* base_data = base_elements.data();
+    for (size_t i = 0; i < target_size; i++) {
+        CppType sum = 0;
+        CppType base_sum = 0;
+        CppType target_sum = 0;
+        size_t dim_size = target_offset[i + 1] - target_offset[i];
+        CppType result_value = 0;
+        for (size_t j = 0; j < dim_size; j++) {
+            sum += base_data[j] * target_data[j];
+            if constexpr (!isNorm) {
+                base_sum += base_data[j] * base_data[j];
+                target_sum += target_data[j] * target_data[j];
+            }
+        }
+        if constexpr (!isNorm) {
+            result_value = sum / (std::sqrt(base_sum) * std::sqrt(target_sum));
```

Comment:
Is it possible that `(std::sqrt(base_sum) * std::sqrt(target_sum))` is zero?

Reply:
If fe doesn't optimize the null case to be a constant, it could be 0, so just in case, I'm going to add a non-0 check here.